system_metrics_collector/test/moving_average_statistics/test_moving_average_statistics.cpp

```diff
@@ -0,0 +1,194 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <thread>
+#include <atomic>
+#include "moving_average_statistics/moving_average.hpp"
+
+
+/**
+ * Test fixture
+ */
+class MovingAverageStatisticsTestFixture : public ::testing::Test
+{
+public:
+  void SetUp() override
+  {
+    moving_average_statistics = std::make_shared<MovingAverageStatistics<double>>();
+  }
+
+  void TearDown() override
+  {
+    moving_average_statistics->reset();
+  }
+
+protected:
+  std::shared_ptr<MovingAverageStatistics<double>> moving_average_statistics = nullptr;
+  int expected_count = 0;
+};
+
+// Useful testing constants
+const double TEST_DATA[] = {-3.5, -2.1, -1.1, 0.0, 4.7, 5.0, 6.7, 9.9, 11.0};
+const double EXPECTED_AVG = 3.4;
+const double EXPECTED_MIN = -3.5;
+const double EXPECTED_MAX = 11.0;
+const double EXPECTED_STD = 4.997999599839919955173;
+const double EXPECTED_SIZE = 9;
+
+TEST_F(MovingAverageStatisticsTestFixture, sanity) {
+  ASSERT_TRUE(true);
+  ASSERT_NE(moving_average_statistics, nullptr);
+}
+
+TEST_F(MovingAverageStatisticsTestFixture, test_average) {
+  for (double d : TEST_DATA) {
+    moving_average_statistics->add_item(d);
+    ASSERT_EQ(++expected_count, moving_average_statistics->get_count());
+  }
+
+  EXPECT_DOUBLE_EQ(moving_average_statistics->average(), EXPECTED_AVG);
+}
+
+TEST_F(MovingAverageStatisticsTestFixture, test_average_empty) {
+  ASSERT_TRUE(std::isnan(moving_average_statistics->average()));
+}
+
+TEST_F(MovingAverageStatisticsTestFixture, test_maximum) {
+  for (double d : TEST_DATA) {
+    moving_average_statistics->add_item(d);
+    ASSERT_EQ(++expected_count, moving_average_statistics->get_count());
+  }
+
+  EXPECT_EQ(moving_average_statistics->max(), EXPECTED_MAX);
+}
+
+TEST_F(MovingAverageStatisticsTestFixture, test_maximum_empty) {
+  ASSERT_TRUE(std::isnan(moving_average_statistics->max()));
+}
+
+TEST_F(MovingAverageStatisticsTestFixture, test_minimum) {
+  for (double d : TEST_DATA) {
+    moving_average_statistics->add_item(d);
+    ASSERT_EQ(++expected_count, moving_average_statistics->get_count());
+  }
+
+  const double expected_minimum = -3.5;
+  EXPECT_EQ(moving_average_statistics->min(), expected_minimum);
+}
+
+TEST_F(MovingAverageStatisticsTestFixture, test_minimum_empty) {
+  ASSERT_TRUE(std::isnan(moving_average_statistics->min()));
+}
+
+TEST_F(MovingAverageStatisticsTestFixture, test_standard_deviation) {
+  for (double d : TEST_DATA) {
+    moving_average_statistics->add_item(d);
+  }
```

Comment:
This code is duplicated multiple times. Maybe a utility function would be better?

Reply:
Will do.